File: baekjoon/src/tags/dynamic/No1966.java
package tags.dynamic;

import java.io.*;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/1966
 */
public class No1966 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static StringTokenizer tokenizer;

    public static int[][] houses;

    public static void setValues() throws IOException {
        int numberOfInput = Integer.parseInt(reader.readLine());
        houses = new int[numberOfInput][3];
        for (int i = 0; i < numberOfInput; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int r = Integer.parseInt(tokenizer.nextToken());
            int g = Integer.parseInt(tokenizer.nextToken());
            int b = Integer.parseInt(tokenizer.nextToken());
            houses[i] = new int[]{r, g, b};
        }
    }

    public static int getCost() {
        int length = houses.length;
        for (int i = 1; i < length; i++) {
            houses[i][0] += Math.min(houses[i - 1][1], houses[i - 1][2]);
            houses[i][1] += Math.min(houses[i - 1][0], houses[i - 1][2]);
            houses[i][2] += Math.min(houses[i - 1][0], houses[i - 1][1]);
        }
        return Math.min(Math.min(houses[length - 1][0], houses[length - 1][1]), houses[length - 1][2]);
    }

    public static void main(String[] args) throws IOException {
        setValues();
        writer.write(String.valueOf(getCost()));
        writer.flush();
    }
}

File: baekjoon/src/tags/data_structures/No1966.kt
package tags.data_structures

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.LinkedList

/**
 * https://www.acmicpc.net/problem/1966
 */
fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))

    val numberOfInput = reader.readLine().toInt()
    repeat(numberOfInput) {
        val target = reader.readLine().split(" ").map { it.toInt() }[1]
        val priorities = reader.readLine().split(" ").map { it.toInt() }
        writer.write("${getOrder(target, priorities)}")
        writer.newLine()
    }
    writer.flush()
}

private fun getOrder(target: Int, priorities: List<Int>): Int {
    val queue = LinkedList(priorities)
    var indexOfTarget = target

    val sortedPriorities = priorities.sortedDescending()
    var priorityIndex = 0

    while (true) {
        val currentPriority = queue.pollFirst()
        if (sortedPriorities[priorityIndex] == currentPriority) {
            if (indexOfTarget == 0) {
                priorityIndex++
                break
            }
            priorityIndex++
        } else
            queue.addLast(currentPriority)
        indexOfTarget = if (indexOfTarget == 0) queue.size - 1 else indexOfTarget - 1
    }
    return priorityIndex
}
